### DISCUSSION TITLE
Null-Removing Function

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5980] [Q&A] [answered]
I would like to create an Enumerable extension method, ```WhereIsNotNull```, that will filter an enumerable to contain only non-null entries and return an enumerable with a non-null type.

Ie.:
```
WhereIsNotNull( IEnumerable<Person?> ) => IEnumerable<Person>
WhereIsNotNull( IEnumerable<int?> ) => IEnumerable<int>
```

Is there a way to declare constraints in such a way to enable this (ie, ```where T : nullable```) or will I need two implementations: one for classes and one for structs?

**Comment by 333fred** [comment, marked as answer]
You will need two implementations.

**Reply by FaustVX** [reply]
[sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgBgAJVMAWAbj0JMwDoAZASygEcrc8BiKCAGz4BDAEZ8ApkTFQR46gGZaANhLoiAYSIBvPEV1EADmCYA3QcAmllqMkQCyglgAoAlDr3bcpsEQDORALxEUGIA7kQswAD8ANoAuppEBDBEmMm8AkQAvnQA6gAWYmBiAHIA9sDF/HwujII+wC4ces1EXkQAxgFBoUSlwgBWYu1RcQlJKWlVWbkFRWUVVTUMdQ3OTXqZbrpbJAqWtHIAPAAqAHxE+YUl5ZUCJ6eOwHlMfqRHx5HnPqUQYO1irlwLT0IVmEmORBAKh2zX8Xx+fzEMyujn2THMAFsAud0WIsS8guUglVnABCdbbfB7TBWTDvc6XOY3Kr3R7PV50k6fXwI/6A5qgq5ECFQ+qQYY7OE837/ZFFVE08KY7FKvGkugACTqADVBHwIACYXo6ABlMTiYYK5S4rFSm3q3X6gEcTJAA==)

**Reply by TahirAhmadov** [reply]
This is yet another request from a user which is caused by the "defaultability" approach (instead of "nullability") for unconstrained generics. No need to resume that debate - but I do have to note this is the 3rd or 4th such question that was raised recently.

**Reply by 333fred** [reply]
> No need to resume that debate - but I do have to note this is the 3rd or 4th such question that was raised recently.

I'll be honest Tahir: this wording implies that you want to resume the debate. We get it. No one on the ldm has said we would have designed nullability this way if we were restarting the language. But we aren't restarting the language, and subtle digs like this just make me less inclined to engage with you in other issues.

**Reply by TahirAhmadov** [reply]
No I don't. I wrote everything there was to write about it. It wasn't meant as a subtle dig, either - just pointing something out. 
Again, I'm not going to continue that topic here or anywhere else.

**Reply by TonyValenti** [reply]
I'm curious - although it may not be possible with C#, using IL, would that be possible?

**Reply by HaloFour** [reply]
@TonyValenti 

> I'm curious - although it may not be possible with C#, using IL, would that be possible?

No, the limitation originates at the bytecode and metadata level.  The problem is that with NRTs `IEnumerable<T?>` is `IEnumerable<T>` and with NVTs `IEnumerable<T?>` is really `IEnumerable<Nullable<T>>` so you end up with incompatible method signatures and require different IL to work with, and generics in the runtime is not designed to work like that.

**Reply by TonyValenti** [reply]
@HaloFour Thanks for the clarification.  That's what I thought but I wasn't sure.
